### ISSUE TITLE
[REVIEW]: pylj: A teaching tool for classical atomistic simulation

### ISSUE THREAD

**Issue by whedon** [openjournals/jose-reviews#19]
**Submitting author:** @arm61 (<a href="http://orcid.org/0000-0003-3381-5911">Andrew McCluskey</a>)
**Repository:** <a href="https://github.com/arm61/pylj" target ="_blank">https://github.com/arm61/pylj</a>
**Version:** v1.0.0
**Editor:** @labarba
**Reviewer:** @shivupa, @TJFord
**Archive:** <a href="https://doi.org/10.5281/zenodo.1312617" target="_blank">10.5281/zenodo.1312617</a>

## Status

[![status](http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d/status.svg)](http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d)

Status badge code:

```
HTML: <a href="http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d"><img src="http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d/status.svg"></a>
Markdown: [![status](http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d/status.svg)](http://jose.theoj.org/papers/58daa1a1a564dc8e0f99ffcdae20eb1d)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/arm61/pylj" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@shivupa & @TJFord, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/jose-reviews/invitations

The reviewer guidelines are available here: https://jose.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @labarba know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/arm61/pylj">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0.0)?
- [x] **Authorship:** Has the submitting author (@arm61) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/arm61/pylj">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0.0)?
- [x] **Authorship:** Has the submitting author (@arm61) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @shivupa, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/jose-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/jose-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/jose-papers/blob/jose.00019/jose.00019/10.21105.jose.00019.pdf)

**Comment by shivupa** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Open the review issue
@whedon start review

🚧 🚧 🚧 Experimental Whedon features 🚧 🚧 🚧

# Compile the paper
@whedon generate pdf

```

**Comment by labarba** [comment]
Isn't our bot `whedon` just the cutest thing?

**Comment by TJFord** [comment]
A few comments:
1.  I really like the idea of running a demo through Jupyter, as it is convenient to test the code directly. I tried the cloud version for molecular_dynamics but not successful. Please correct me if I did it wrong. http://35.230.133.1/notebook/notebooks/examples/molecular_dynamics.ipynb#
The error message is here: 

<pre>TypeError                                 Traceback (most recent call last)
<ipython-input-17-980173a92a59> in <module>()
----> 1 system = md_simulation(100, 273.15, 100, 5000, 50)
<ipython-input-16-0578d5628985> in md_simulation(number_of_particles, temperature, box_length, number_of_steps, sample_frequency)
     16                                                                                                  system.cut_off)
     17         # Run the equations of motion integrator algorithm
---> 18         system.particles = md.velocity_verlet(system.particles, system.timestep_length, system.box_length)
     19         # Sample the thermodynamic and structural parameters of the system
     20         system = md.sample(system.particles, system.box_length, system.initial_particles, system)
TypeError: velocity_verlet() missing 1 required positional argument: 'cut_off' </pre>

2. I tried to run it on a linux computer through a terminal, it was not successful either. I followed exactly all the steps shown in README.md.  I suspected I missed installing some libraries. I think it will be beneficial if a detailed installation manual can be provided.

**Comment by shivupa** [comment]
Can you open this as an issue on the submission repo. I had some questions about this too.

**Comment by TJFord** [comment]
@shivupa I just opened an issue at https://github.com/arm61/pylj/issues/11

**Comment by labarba** [comment]
@shivupa, @TJFord -- the submitting author is also working on improved narrated examples for this package. We will wait for a heads-up that those are ready, and then please have a look at them, too.

**Comment by TJFord** [comment]
@labarba  @shivupa it looks like the example monte carlo simulation is not working either. @arm61, can you double check it? 
Also a few other minor suggestions: 
1. would it be better if the title is changed to "pylj: A teaching tool for classical molecular dynamics simulation "? as the focus of the tool is on molecular dynamics simulation. 
2. I would like to cite a few other widely used open source molecular dynamics simulation packages in the paper, such as lammps, gromacs, etc., for those students if they want to know more about it. 
 
<pre>---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-22-4859d5764793> in <module>()
----> 1 system = mc_simulation(2, 273.15, 45, 100000, 100)

<ipython-input-21-456fcaae123c> in mc_simulation(number_of_particles, temperature, box_length, number_of_steps, sample_frequency)
      8     system.particles, system.distances, system.energies = comp.compute_energy(system.particles, 
      9                                                                               system.box_length,
---> 10                                                                               system.cut_off)
     11     old_energy = system.energies.sum()
     12     system = mc.sample(system.particles, system.box_length, old_energy, system)

AttributeError: 'System' object has no attribute 'cut_off'</pre>

**Comment by labarba** [comment]
I agree that it should say "molecular simulation," or better "molecular dynamics simulation," everywhere.

Folks from outside the MD word will have no idea what the author means otherwise.

**Comment by shivupa** [comment]
I'd suggest "molecular simulation" since this encompasses molecular dynamics and Monte Carlo.

**Comment by arm61** [comment]
Can we possibly agree on "classical atomistic simulation"? For the following reasons:

- Since we are only really dealing with individual atoms of argon in pylj it is a bit false to say molecular.
- I would like to keep "classical" in there as pylj only performs classical potential based simulation, e.g. there are not quantum mechanical methods (e.g. DTF) to be seen.

**Comment by arm61** [comment]
Paper title has been updated in https://github.com/arm61/pylj/commit/bb492bed8d8d37656e2534574f8ba4a876d001a6

**Comment by arm61** [comment]
There are now two new examples of possible exercises showing the utility of pylj these can be found https://github.com/arm61/pylj/tree/master/examples . This means there is now the molecular dynamic and monte carlo examples showing how pylj can be used for teaching about simulation methods and the ideal gas example showing an application of pylj to build on traditional physical chemistry lecture material. 

Note that to run these you need to build the most recent commit of pylj (this is because in developing the tutorials I recognised some more useful pedagogical methods that could be applied). This has also lead to the pairwise module which is a pure python implementation of the comp module (this is MUCH slower than the cython but possibly more pedagogically interesting). 

I hope this takes the submission up to the expected level that is required. Sorry about the delay (life gets in the way sometimes).

**Comment by labarba** [comment]
I have an observation that is optional for you to address, but I ask you to consider… You are using British spelling in the `mc.initialise()` function. This is likely to trip many if not most users! I see that _you_ are based in Bath, so it make sense for your local users. Maybe you can overload the name and have two functions that do the same thing?

**Comment by arm61** [comment]
@TJFord I have added references to the paper to gromacs, lammps and dlpoly in https://github.com/arm61/pylj/commit/21baa244c887c5ced60d227ecd0bc30967ac104e

**Comment by arm61** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/jose-papers/blob/jose.00019/jose.00019/10.21105.jose.00019.pdf)

**Comment by TJFord** [comment]
@labarba I think the author @arm61 has addressed all my concerns. I suggest that it is ready for publication.

**Comment by shivupa** [comment]
@labarba I also agree @arm61  has addressed all concerns and it is ready for publication.

**Comment by shivupa** [comment]
@TJFord I'm unsure how whedon works, but we may need to check all the boxes above to proceed.

**Comment by arm61** [comment]
@labarba it seems that both reviewers are happy. I was wondering what the next stage was?

**Comment by arm61** [comment]
also a massive thanks to @shivupa and and @TJFord for all their hard work on making pylj a MUCH more useful piece of software than previously. Great peer reviewing folks :smile:

**Comment by labarba** [comment]
Please stay tuned. I am at a conference now. I have to go over the comments/improvements and paper myself before accepting the submission.

**Comment by labarba** [comment]
@arm61 Could you try moving the positioning of your figure down, so that more text moves to the first page, under Summary? There's a big chunk of white space there.

**Comment by labarba** [comment]
*Editorial feedback*

_Summary: _
(l.1) classical simulation—> classical atomistic simulation
(l.2) use bibles for the citation (and elsewhere)
(l.4) define NVE and NVT
(l.5) capitalize Argon
(l.7) capitalize Matplotlib
 (l.9) explaination—>explanation (typo)
This whole passage is a little hard to parse. The construction is perhaps a bit ritzy. Can you simplify?Something like:
PyLJ is written in Python (using Cython for pairwise interactions) and uses Jupyter notebooks and Matplotlib for visualization (example below). It can be easily deployed in a computer laboratory, and students interact with it without needing to use the command line, like they would when using other molecular dynamics software like Gromacs, DL_poly or others. (Refs.) We provide example notebooks in the repository, showing how to use PyLJ to simulate a 2D gas system using molecular dynamics and Monte-Carlo methods. A variety of other applications are possible.
(p.2) “thermostating” is jargon: please define or rephrase

_Statement of need:_
classical simulation—> classical atomistic simulation
PyLJ allows this by offering—> PyLJ offers
capitalize Matplotlib
“In addition to use in the introduction of the simulation itself” —> huh? Please rephrase.
—.
Audience: you mention in the first sentence undergraduates in chemistry and physics. Can you say more about where in the curriculum this would be used? (what course? what year?)

**Comment by labarba** [comment]
Sorry to give you comments piecemeal… I did the first part on a flight yesterday. 
 
You give a short description of the examples, with a link to the folder in the repo, within the “Statement of Need” section. It doesn’t quite belong there. You could use a new heading “Usage” and talk about the examples there. It would also be good to add a bulleted list of the examples in the README. Navigating down the folders to find the examples is not user-friendly, and since some folders have two notebooks, a user may not know where to start. I, myself, click through the various notebooks, and don’t know what order one is supposed to go through this. 

What I’d like to see is a paragraph in the paper (and maybe in a new README in the `examples` folder) that orients an independent learner or instructor on getting started and adopting this tool. It took me _many_ visits to your repository to finally stumble on the `guide.ipynb` file just now! Please add a link to this file from the README. I think the “How to teach with PyLJ” section here would be useful to have in the paper. (But please don’t use the un-semantic hyperlinked “here.”)

**Comment by labarba** [comment]
I went to the [Getting Started](http://pythoninchemistry.org/getting-started/) page on the documentation, and clicked on the Tutorial links, which launch a Binder session. However, they all give a "cannot resolve ref" error.

**Comment by arm61** [comment]
whedon generate pdf

**Comment by arm61** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/jose-papers/blob/jose.00019/jose.00019/10.21105.jose.00019.pdf)

**Comment by labarba** [comment]
Maybe the problem with the citation to [LENNARD-JONES](https://github.com/arm61/pylj/blob/8d6ebd3a02049f2290249648e4a988ac7d82ad0e/paper/paper.bib#L57) is that the entry has no author field.

**Comment by arm61** [comment]
Okay, I think that is everything that you've mentioned resolved. Hopefully, the paper is more suitable not (commits https://github.com/arm61/pylj/commit/8d6ebd3a02049f2290249648e4a988ac7d82ad0e https://github.com/arm61/pylj/commit/94a5a49d57cb76e94508bb40b2ee3300e5b03ed1 https://github.com/arm61/pylj/commit/ddf306f3896bc02d6b18555bfa01c1ee0844daa4 https://github.com/arm61/pylj/commit/22b926f1ecfc26b95269cdefd98fe3e427dafdb2). The only thing that you suggested that I haven't done is captalising argon, this is because (as I understand) it is IUPAC convention to not trade element names as proper nouns (https://chemistry.stackexchange.com/questions/6381/why-do-people-often-capitalize-element-names). 

I have also added a mention of the examples in the README (commit https://github.com/arm61/pylj/commit/666951580418427cc863f7028a39547b5794e8be). I don't go into detail as my plan is to mention the paper once it is published which gives more details. 

I have move the "How to teach with pylj" bit from the guide.ipynb file to the "Usage" section of the paper and as a result have decided that the guide.ipynb file is surplus to requirements. 

Thanks for the thorough comments @labarba

**Comment by arm61** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/jose-papers/blob/jose.00019/jose.00019/10.21105.jose.00019.pdf)

**Comment by arm61** [comment]
Done! https://github.com/arm61/pylj/commit/d5f0a23e03c0eb06657db96a9f7c299a4a932902

**Comment by arm61** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/jose-papers/blob/jose.00019/jose.00019/10.21105.jose.00019.pdf)

**Comment by labarba** [comment]
OK! Now, please make a deposit on Zenodo or your favorite archive (or update a versioned DOI, if the case), and give me the DOI here.

**Comment by arm61** [comment]
The DOI is 10.5281/zenodo.1312617

**Comment by labarba** [comment]
@whedon set 10.5281/zenodo.1312617 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1312617" target="_blank">10.5281/zenodo.1312617</a> is the archive.

**Comment by labarba** [comment]
I need to do a little local processing to publish your paper, but will have to do that tomorrow, due to commitments now.

**Comment by labarba** [comment]
Something is wrong with the Zenodo DOI you gave me
https://doi.org/10.5281/zenodo.1312617

"This DOI cannot be found in the DOI System"

Can you double check?

**Comment by labarba** [comment]
OK. We'll wait. If the DOI does not resolve tomorrow morning your time (please check), then send an email to Zenodo help. I won't publish the JOSE paper until the DOI of the archive resolves.

**Comment by labarba** [comment]
I mean, at the moment it _is_ pushed to [JOSE-papers](https://github.com/openjournals/jose-papers), but I haven't made a Crossref deposit.

**Comment by labarba** [comment]
I checked and the DOI still gives me an error: https://doi.org/10.5281/zenodo.1312617

**Comment by arm61** [comment]
Same, I have contacted Zenodo via both twitter and email to see what the issue is

**Comment by arm61** [comment]
It is resolved now

**Comment by labarba** [comment]
Crossref threw citation errors. Please fix DOI links for the Gromacs paper and Plimpton 1995.

I will have to do the whole process again!

**Comment by arm61** [comment]
Resolved in https://github.com/arm61/pylj/commit/ce9168990facc09b3f2441e342902d135c2bae7e

Sorry about that, apparently the bibtex served by sciencedirect incorrectly prepends "http://doi.org/" for the doi...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://jose.theoj.org/papers/10.21105/jose.00019/status.svg)](https://doi.org/10.21105/jose.00019)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/jose.00019">
  <img src="https://jose.theoj.org/papers/10.21105/jose.00019/status.svg" alt="DOI badge" >
</a>
```
This is how it will look in your documentation:

[![DOI](https://jose.theoj.org/papers/10.21105/jose.00019/status.svg)](https://doi.org/10.21105/jose.00019)

**We need your help!**

Journal of Open Source Education is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. 
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3ODA=

**Comment by labarba** [comment]
@arm61  -- your paper is now published in JOSE ‼️  Congratulations 🎉 
You can now add the happy green JOSE button on your repository.

@shivupa, @TJFord — Thank you again for your review 🙏

**Comment by arm61** [comment]
Yay! Thanks @labarba for your awesome editoral work, and thanks again to @shivupa and @TJFord!

**Comment by shivupa** [comment]
Congrats! I'm happy to have been a part of this.

**Comment by labarba** [comment]
:loudspeaker:  New Reviewer Sign-up Form!
https://forms.gle/Rizd3TcHnQKhrbYY7
